examples/hello/lib/hello/app.rb

```diff
@@ -0,0 +1,66 @@
+require "hello/actions"
+
+module Hello
+  class App
+    class Router
+      class NotFound < StandardError; end
+
+      attr_reader :routes
+
+      def initialize(&block)
+        @routes = {}
+        instance_eval(&block)
+      end
+
+      def resource(name, at:, &block)
+        @this_path = at
+        routes[@this_path] = {}
+        instance_eval(&block)
+      end
+
+      def on(req_method, call:)
+        routes[@this_path][req_method] = call
+      end
+
+      def action_for(path, req_method)
+        routes.fetch(path).fetch(req_method)
+      rescue KeyError
+        raise NotFound.new("Can't find route for #{req_method} #{path}")
+      end
+    end
+
+    def call(env)
+      router = Router.new do
+        resource "hello", at: "/hello" do
+          on "GET", call: "say_hello"
+          on "HEAD", call: "do_nothing"
+        end
+
+        resource "world", at: "/world" do
+          on "DELETE", call: "say_goodbye"
+        end
+      end
+
+      path = env.fetch("PATH_INFO")
+      req_method = env.fetch("REQUEST_METHOD")
+
+      begin
+        action_name = router.action_for(path, req_method)
+        action = actions.send(action_name)
+        body = action.call
```

Comment:
I'll try to explain where it's at now and where I'd like to end up. That will help:

An action is just an object with a `call` method. It should encapsulate some use case of the system. For example:

```ruby
action = CreateArticle.new
result = action.call(data)
```

The `actions` object here is a factory object that returns action objects. The router as it is now can take a request and return the *name* of an action (so perhaps `action_for` is misleading). The app object is then sending that action name to the actions factory to get an action object, and then calling the action.

The result the action is just another object, but in this very super early stage I'm treating that object as the intended responsej body. Eventually there will be a more explicit way to turn the result of an action into an HTTP response. See #4 

What I don't like is the part where my router returns a string representing the action name, which I then have to use to get the actual action object. I could make my router depend on the actions factory, but I really wanted it to be a pure-configuration object (maybe even explicitly immutable). My vision for this is that if you need routing anywhere, you can easily pass in an instance of the router on the cheap. I can see several actions depending on the router, so I would prefer the router to not depend on the actions. This is where I'm scratching my head.

I haven't had time to keep driving this with more tests, but I'm hoping once I do the path will be clearer.